dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -0,0 +1,421 @@
+    // - Icon: Retrieve from Shell. If it's not the case or the file extension is unsupported, then throw.
+    winrt::com_ptr<IPropertyStore> propertyStore;
+    wil::unique_hwnd hWindow{ GetConsoleWindow() };
+    THROW_IF_FAILED(SHGetPropertyStoreForWindow(hWindow.get(), IID_PPV_ARGS(propertyStore.put())));
+
+    // Retrieve the display name
+    std::wstring displayName{};
+
+    wil::unique_prop_variant propVariantDisplayName;
+    // Do not throw in case of failure. We can use the filepath approach below as fallback to set a DisplayName.
+    LOG_IF_FAILED(propertyStore->GetValue(PKEY_AppUserModel_RelaunchDisplayNameResource, &propVariantDisplayName));
+
+    displayName = propVariantDisplayName.vt != VT_EMPTY ? propVariantDisplayName.pwszVal : L"";
+
+    if (displayName.length() == 0)
+    {
+        wil::unique_cotaskmem_string processName;
+        THROW_IF_FAILED(GetCurrentProcessPath(processName));
+
+        displayName = processName.get();
+
+        size_t lastBackslashPosition{ displayName.rfind(L"\\") };
+        THROW_HR_IF(E_UNEXPECTED, lastBackslashPosition == std::wstring::npos);
+        displayName = displayName.substr(lastBackslashPosition + 1); // One after the delimiter
+
+        displayName.erase(displayName.find_first_of(L".")); // Remove file extension
+    }
+
+    // Retrieve the icon
+    std::wstring iconFilePath{};
+
+    wil::unique_prop_variant propVariantIcon;
+    // Throw in case of failure, since Icon is mandatory and we don't have a fallback!
+    THROW_IF_FAILED(propertyStore->GetValue(PKEY_AppUserModel_RelaunchIconResource, &propVariantIcon));
+    THROW_HR_IF_MSG(E_UNEXPECTED, propVariantIcon.vt == VT_EMPTY, "You must specify an app icon before calling Register().");
+
+    iconFilePath = propVariantIcon.pwszVal;
+
+    // Icon filepaths from Shell APIs have this format: <filepath>,-<index>,
+    // since .ico files can have multiple icons in the same file.
+    // NotificationController doesn't seem to support such format, so let it take the first icon by default.
+    auto iteratorForCommaDelimiter = iconFilePath.find_first_of(L",");
+    if (iteratorForCommaDelimiter != std::wstring::npos) // It may or may not have an index, which is fine.
+    {
+        iconFilePath.erase(iteratorForCommaDelimiter);
+    }
+
+    auto iteratorForFileExtension = iconFilePath.find_first_of(L".");
+    THROW_HR_IF_MSG(E_UNEXPECTED, iteratorForFileExtension == std::wstring::npos, "You must provide a valid filepath as the app icon.");
+
+    std::wstring iconFileExtension = iconFilePath.substr(iteratorForFileExtension);
+    THROW_HR_IF_MSG(E_UNEXPECTED, iconFileExtension != L".ico" && iconFileExtension != L".png",
+        "You must provide a supported file extension as the icon (.ico or .png).");
+
+    RegisterValue(hKey, L"DisplayName", reinterpret_cast<const BYTE*>(displayName.c_str()), REG_EXPAND_SZ, displayName.size() * sizeof(wchar_t));
```

Comment:
>RegisterValue [](http://example.com/codeflow?start=4&length=13)

Future: The signature of this method seems strange. I would have expected:

RegisterValue(HKEY key, const wchar_t* name, std::wstring_view value, bool expandString);

... and then it's up to RegisterValue to compute the lengths of the incoming thing.

Reply:
This is good improvement to add :)

Filing issue: https://github.com/microsoft/WindowsAppSDK/issues/2222